src/aws-cpp-sdk-core/include/smithy/tracing/TracingUtils.h

```diff
@@ -0,0 +1,97 @@
+/**
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ * SPDX-License-Identifier: Apache-2.0.
+ */
+#pragma once
+
+#include <aws/core/monitoring/HttpClientMetrics.h>
+#include <smithy/Smithy_EXPORTS.h>
+#include <smithy/tracing/Meter.h>
+#include <functional>
+#include <chrono>
+#include <utility>
+
+namespace smithy {
+    namespace components {
+        namespace tracing {
+            static const char SMITHY_METRICS_DNS_DURATION[] = "smithy.client.http.dns_duration";
+            static const char SMITHY_METRICS_CONNECT_DURATION[] = "smithy.client.http.connect_duration";
+            static const char SMITHY_METRICS_SSL_DURATION[] = "smithy.client.http.ssl_duration";
+            static const char SMITHY_METRICS_THROUGHPUT[] = "smithy.client.http.throughput";
+            static const char SMITHY_METRICS_UNKNOWN_METRIC[] = "smithy.client.http.unknown_metric";
+
+            class SMITHY_API TracingUtils {
+            public:
+                TracingUtils() = default;
+
+                template<typename T>
+                static T MakeCallWithTiming(std::function<T()> func,
+                    Aws::String metricName,
+                    std::shared_ptr<Meter> meter,
+                    Aws::Map<Aws::String, Aws::String> attributes,
+                    Aws::String description = "") {
+                    auto before = std::chrono::high_resolution_clock::now();
+                    auto returnValue = func();
+                    auto after = std::chrono::high_resolution_clock::now();
+                    auto duration = std::chrono::duration_cast<std::chrono::milliseconds>(after - before).count();
+                    auto histogram = meter->CreateHistogram(std::move(metricName), "ms", std::move(description));
```

Comment:
please consider CreateHistogram method to return non-ptr type
heap allocation may fail and there is no reliable way to check for this with exceptions disabled and template return value.

Reply:
im not sure we can, because otel returns a unique ptr, so we would have to hold on to that when returned until end of scope if we want to reference it. I can add a nullptr check though, open to other suggestions as well